File: bestia-models/src/main/java/bestia/model/geometry/CollisionHelper.java
package bestia.model.geometry;

/**
 * This class contains the shared collision code for the {@link CollisionShape}
 * implementations. Since collision is implemented with a visitor pattern code
 * would have to be implemented twice. In order to prevent this code sharing all
 * the collision methods are implemented here.
 * 
 * @author Thomas Felix <thomas.felix@tfelix.de>
 *
 */
final class CollisionHelper {

	/**
	 * Priv. ctor. No instance. Use static methods.
	 */
	private CollisionHelper() {
		// no op.
	}

	/**
	 * Checks if a {@link Point} and a {@link Rect} collide.
	 * 
	 * @param s
	 *            Vector.
	 * @param r
	 *            Rect.
	 * @return TRUE if they collide. FALSE otherwise.
	 */
	public static boolean collide(Point s, Rect r) {

		final boolean xLeft = s.getX() < r.getX();
		final boolean yTop = s.getY() < r.getY();
		final boolean xRight = s.getX() > (r.getX() + r.getWidth());
		final boolean yBottom = s.getY() > r.getY() + r.getHeight();

		return !xLeft && !yTop && !xRight && !yBottom;
	}

	/**
	 * Checks if a {@link Circle} and a {@link Rect} collide.
	 * 
	 * @param s
	 *            Circle.
	 * @param r
	 *            Rect.
	 * @return TRUE if they collide. FALSE otherwise.
	 */
	public static boolean collide(Circle s, Rect r) {

		final Point cc = s.getCenter();

		if (r.collide(cc)) {
			return true;
		}

		// Check where the center of the circle is compared to the rectangle.
		final long x = r.getX();
		final long y = r.getY();
		final long x2 = x + r.getWidth();
		final long y2 = y + r.getHeight();

		if (cc.getX() < x && cc.getY() < y) {
			// Top left case.

			final int d = (int) Math.sqrt((cc.getX() - x) * (cc.getX() - x) +
					(cc.getY() - y) * (cc.getY() - y));
			return d <= s.getRadius();
		} else if (cc.getX() >= x && cc.getX() <= x2 && cc.getY() <= y) {
			// Top case.
			final long d = y - cc.getY();
			return d <= s.getRadius();
		} else if (cc.getX() > x2 && cc.getY() < y) {
			// Right top case.
			final int d = (int) Math.sqrt((x2 - cc.getX()) * (x2 - cc.getX()) + (y - cc.getY()) * (y - cc.getY()));
			return d <= s.getRadius();
		} else if (cc.getX() >= x2 && cc.getY() >= y && cc.getY() <= y2) {
			// Right case.
			final long d = cc.getX() - x2;
			return d <= s.getRadius();
		} else if (cc.getX() > x2 && cc.getY() > y2) {
			// Right bottom case.
			final int d = (int) Math.sqrt((cc.getX() - x2) * (cc.getX() - x2) +
					(cc.getY() - y2) * (cc.getY() - y2));
			return d <= s.getRadius();
		} else if (cc.getX() >= x && cc.getX() <= x2 && cc.getY() >= y2) {
			// Bottom case.
			final long d = cc.getY() - y2;
			return d <= s.getRadius();
		} else if (cc.getX() < x && cc.getY() > y2) {
			// Bottom left case.
			final int d = (int) Math.sqrt((cc.getX() - x) * (cc.getX() - x) + (cc.getY() - y2) * (cc.getY() - y2));
			return d <= s.getRadius();
		} else {
			// Left case.
			final long d = x - cc.getX();
			return d <= s.getRadius();
		}
	}

	/**
	 * Checks if a {@link Circle} and a {@link Point} collide.
	 * 
	 * @param c
	 *            Circle.
	 * @param v
	 *            Vector.
	 * @return TRUE if they collide. FALSE otherwise.
	 */
	public static boolean collide(Circle c, Point v) {
		final long distance = Math.abs(c.getCenter().getX() - v.getX()) +
				Math.abs(c.getCenter().getY() - v.getY());
		return (distance <= c.getRadius());
	}

	/**
	 * Checks if two {@link Rect} collides.
	 * 
	 * @param r1
	 *            First Rect.
	 * @param r2
	 *            Second Rect.
	 * @return TRUE if they collide. FALSE otherwise.
	 */
	public static boolean collide(Rect r1, Rect r2) {
		final boolean xCheck1 = r1.getX() < r2.getX() + r2.getWidth();
		final boolean xCheck2 = r1.getX() + r1.getWidth() > r2.getX();
		final boolean yCheck1 = r1.getY() < r2.getY() + r2.getHeight();
		final boolean yCheck2 = r1.getHeight() + r1.getY() > r2.getY();

		return xCheck1 && xCheck2 && yCheck1 && yCheck2;
	}

	/**
	 * Checks if two {@link Circle} collide.
	 * 
	 * @param s
	 *            First circle.
	 * @param s2
	 *            Second circle.
	 * @return TRUE if they collide. FALSE otherwise.
	 */
	public static boolean collide(Circle s, Circle s2) {
		final double distance = s.getCenter().getDistance(s2.getCenter());
		return (distance < s.getRadius() + s2.getRadius());
	}

	/**
	 * Checks if two {@link Point} collide.
	 * 
	 * @param s
	 *            First vector.
	 * @param s2
	 *            Second vector.
	 * @return TRUE if they collide. False otherwise.
	 */
	public static boolean collide(Point s, Point s2) {
		return s.equals(s2);
	}

}

File: bestia-models/src/main/kotlin/net/bestia/model/geometry/CollisionHelper.kt
package net.bestia.model.geometry

/**
 * This class contains the shared collision code for the [Shape]
 * implementations. Since collision is implemented with a visitor pattern code
 * would have to be implemented twice. In order to prevent this code sharing all
 * the collision methods are implemented here.
 *
 * @author Thomas Felix
 */
internal object CollisionHelper {

  /**
   * Checks if a [Point] and a [Rect] collide.
   *
   * @param s
   * Vector.
   * @param r
   * Rect.
   * @return TRUE if they collide. FALSE otherwise.
   */
  fun collide(s: Point, r: Rect): Boolean {

    val xLeft = s.x < r.x
    val yTop = s.y < r.y
    val xRight = s.x > r.x + r.width
    val yBottom = s.y > r.y + r.height

    return !xLeft && !yTop && !xRight && !yBottom
  }

  /**
   * Checks if a [Circle] and a [Rect] collide.
   *
   * @param s
   * Circle.
   * @param r
   * Rect.
   * @return TRUE if they collide. FALSE otherwise.
   */
  fun collide(s: Circle, r: Rect): Boolean {

    val cc = s.center

    if (r.collide(cc)) {
      return true
    }

    // Check where the center of the circle is compared to the rectangle.
    val x = r.x
    val y = r.y
    val x2 = x + r.width
    val y2 = y + r.height

    if (cc.x < x && cc.y < y) {
      // Top left case.

      val d = Math.sqrt(((cc.x - x) * (cc.x - x) + (cc.y - y) * (cc.y - y)).toDouble()).toInt()
      return d <= s.radius
    } else if (cc.x in x..x2 && cc.y <= y) {
      // Top case.
      val d = y - cc.y
      return d <= s.radius
    } else if (cc.x > x2 && cc.y < y) {
      // Right top case.
      val d = Math.sqrt(((x2 - cc.x) * (x2 - cc.x) + (y - cc.y) * (y - cc.y)).toDouble()).toInt()
      return d <= s.radius
    } else if (cc.x >= x2 && cc.y >= y && cc.y <= y2) {
      // Right case.
      val d = cc.x - x2
      return d <= s.radius
    } else if (cc.x > x2 && cc.y > y2) {
      // Right bottom case.
      val d = Math.sqrt(((cc.x - x2) * (cc.x - x2) + (cc.y - y2) * (cc.y - y2)).toDouble()).toInt()
      return d <= s.radius
    } else if (cc.x in x..x2 && cc.y >= y2) {
      // Bottom case.
      val d = cc.y - y2
      return d <= s.radius
    } else if (cc.x < x && cc.y > y2) {
      // Bottom left case.
      val d = Math.sqrt(((cc.x - x) * (cc.x - x) + (cc.y - y2) * (cc.y - y2)).toDouble()).toInt()
      return d <= s.radius
    } else {
      // Left case.
      val d = x - cc.x
      return d <= s.radius
    }
  }

  /**
   * Checks if a [Circle] and a [Point] collide.
   *
   * @param c
   * Circle.
   * @param v
   * Vector.
   * @return TRUE if they collide. FALSE otherwise.
   */
  fun collide(c: Circle, v: Point): Boolean {
    val distance = Math.abs(c.center.x - v.x) + Math.abs(c.center.y - v.y)
    return distance <= c.radius
  }

  /**
   * Checks if two [Rect] collides.
   *
   * @param r1
   * First Rect.
   * @param r2
   * Second Rect.
   * @return TRUE if they collide. FALSE otherwise.
   */
  fun collide(r1: Rect, r2: Rect): Boolean {
    val xCheck1 = r1.x < r2.x + r2.width
    val xCheck2 = r1.x + r1.width > r2.x
    val yCheck1 = r1.y < r2.y + r2.height
    val yCheck2 = r1.height + r1.y > r2.y

    return xCheck1 && xCheck2 && yCheck1 && yCheck2
  }

  /**
   * Checks if two [Circle] collide.
   *
   * @param s
   * First circle.
   * @param s2
   * Second circle.
   * @return TRUE if they collide. FALSE otherwise.
   */
  fun collide(s: Circle, s2: Circle): Boolean {
    val distance = s.center.getDistance(s2.center)
    return distance < s.radius + s2.radius
  }

  /**
   * Checks if two [Point] collide.
   *
   * @param s
   * First vector.
   * @param s2
   * Second vector.
   * @return TRUE if they collide. False otherwise.
   */
  fun collide(s: Point, s2: Point): Boolean {
    return s == s2
  }
}
